### PR TITLE
Abstract Cursor Resource Manager

### PR DESCRIPTION
**Pull request by rozza**
Two commits:

1) Auto refactor of the CursorResourceManager - into its own class.
2) Made CursorResourceManager abstract and then implemented it in both sync and async.

You should be able to view both implementations side by side. 

Note, the aim isn't to make the API's / behaviour exactly same - rather its to bring them to a point where they are similar enough, that it isn't such a mental leap to work on them.

### CODE REVIEW

File: driver-core/src/main/com/mongodb/internal/operation/ChangeStreamBatchCursor.java [hunk 1]
```diff
@@ -165,20 +165,20 @@ private void cachePostBatchResumeToken(final AggregateResponseBatchCursor<RawBso
      * @param lastIdConsumer Is {@linkplain Consumer#accept(Object) called} iff {@code rawDocuments} is successfully converted
      *                       and the returned {@link List} is neither {@code null} nor {@linkplain List#isEmpty() empty}.
      */
-    @Nullable
     static <T> List<T> convertAndProduceLastId(@Nullable final List<RawBsonDocument> rawDocuments,
                                                final Decoder<T> decoder,
                                                final Consumer<BsonDocument> lastIdConsumer) {
-        List<T> results = null;
+        List<T> results = new ArrayList<>();
```
Comment (rozza) [inline driver-core/src/main/com/mongodb/internal/operation/ChangeStreamBatchCursor.java:171]: Note: This follows on from the previous AsyncSingleBatchCursor change where `next()` no longer returns a null value.

Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/operation/ChangeStreamBatchCursor.java:171]: Never mind, I mis-read your comment.

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java [hunk 2]
```diff
@@ -102,120 +84,101 @@ class AsyncCommandBatchCursor<T> implements AsyncAggregateResponseBatchCursor<T>
...
-        } else if (initialCommandCursorResult != null && !firstBatchEmpty) {
-            // May be empty for a tailable cursor
-            List<T> results = initialCommandCursorResult.getResults();
-            initialCommandCursorResult = null;
-            if (getServerCursor() == null) {
-                close();
+            callback.onResult(null, new MongoException(MESSAGE_IF_CLOSED_AS_CURSOR));
+            return;
+        }
+
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:118]: Nit: one blank line is enough

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java [hunk 3]
```diff
@@ -102,120 +84,101 @@ class AsyncCommandBatchCursor<T> implements AsyncAggregateResponseBatchCursor<T>
...
+
+
+        resourceManager.execute((AsyncCallbackSupplier<List<T>>) funcCallback -> {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:121]: `funcCallback` is a bit ambiguous.  How about `resourceManagerCallback`?

File: driver-core/src/main/com/mongodb/internal/operation/AsyncOperationHelper.java [hunk 4]
```diff
@@ -61,6 +61,10 @@ interface AsyncCallableWithConnection {
         void call(@Nullable AsyncConnection connection, @Nullable Throwable t);
     }
 
+    interface AsyncCallableConnectionWithCallback<T> {
+        void call(@Nullable AsyncConnection connection, SingleResultCallback<T> callback);
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/operation/AsyncOperationHelper.java:65]: Does this need to be `Nullable`?  I can't think why, since I think the method wouldn't be called if the resource manager fails to get a connection

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java [hunk 5]
```diff
@@ -228,187 +191,154 @@ public int getMaxWireVersion() {
...
+                                    if (!resourceManager.operable()) {
+                                        // The cursor was closed
+                                        resourceManager.releaseServerAndClientResources(connection);
+                                        wrappedCallback.onResult(emptyList(), null);
+                                        return;
+                                    }
+
+                                    List<T> nextBatch = commandCursorResult.getResults();
+                                    if (nextBatch.isEmpty() && commandCursorResult.getServerCursor() != null) {
+                                        getMore(commandCursorResult.getServerCursor(), wrappedCallback);
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:224]: I think this will check out another connection before returning the current one to the pool?                                        // Should it be re-using the same connection(was it doing that previously?)?

**Review by jyemin — CHANGES_REQUESTED** [review]

File: driver-core/src/main/com/mongodb/internal/operation/ChangeStreamBatchCursor.java [hunk 6]
```diff
@@ -165,20 +165,20 @@ private void cachePostBatchResumeToken(final AggregateResponseBatchCursor<RawBso
      * @param lastIdConsumer Is {@linkplain Consumer#accept(Object) called} iff {@code rawDocuments} is successfully converted
      *                       and the returned {@link List} is neither {@code null} nor {@linkplain List#isEmpty() empty}.
      */
-    @Nullable
     static <T> List<T> convertAndProduceLastId(@Nullable final List<RawBsonDocument> rawDocuments,
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/ChangeStreamBatchCursor.java:168]: Yes - as it can be called using `tryNext`

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java (hunk 3 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:119]: `funcCallback` is the existing convention in the codebase where `AsyncCallbackSupplier<>` is used and a variable called `callback` is in scope.

File: driver-core/src/main/com/mongodb/internal/operation/AsyncOperationHelper.java (hunk 4 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncOperationHelper.java:65]: Done

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java [hunk 7]
```diff
@@ -228,187 +191,154 @@ public int getMaxWireVersion() {
...
 
-        int batchSizeForGetMoreCommand = Math.abs(getNumberToReturn(limit, this.batchSize, count.get()));
-        if (batchSizeForGetMoreCommand != 0) {
-            document.append("batchSize", new BsonInt32(batchSizeForGetMoreCommand));
+        ResourceManager(
+                final MongoNamespace namespace,
+                final AsyncConnectionSource connectionSource,
+                @Nullable final AsyncConnection connectionToPin,
+                @Nullable final ServerCursor serverCursor) {
+            super(new ReentrantLock(), namespace, connectionSource, connectionToPin, serverCursor);
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:253]: I'm not 100% sure why two different locks were used / favored originally.

@stIncMale any insights?

File: driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursorHelper.java [hunk 8]
```diff
@@ -65,6 +70,23 @@ static BsonDocument getMoreCommandDocument(
         return document;
     }
 
+    static <T> CommandCursorResult<T> getCommandCursorResult(final ServerAddress serverAddress, final String fieldNameContainingBatch,
+            final BsonDocument commandCursorDocument) {
+        CommandCursorResult<T> commandCursorResult = new CommandCursorResult<>(serverAddress, fieldNameContainingBatch,
+                commandCursorDocument);
+        if (LOGGER.isDebugEnabled()) {
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursorHelper.java:77]: Done

File: driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursor.java [hunk 9]
```diff
@@ -362,79 +311,20 @@ <R> R execute(final String exceptionMessageIfClosed, final Supplier<R> operation
...
-         * demarcated by {@link #tryStartOperation()}/{@link #endOperation()}.
-         */
-        private void doClose() {
             try {
-                if (skipReleasingServerResourcesOnClose) {
-                    serverCursor = null;
-                } else if (serverCursor != null) {
-                    Connection connection = connection();
+                if (getServerCursor() != null) {
+                    // Don't handle corrupted connections
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursor.java:326]: Yes, it makes no sense. Removed. It uses getConnection directly because even if it fails it still needs to do something.

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java (hunk 2 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:118]: Done

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java (hunk 5 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java]: Fixed

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:255]: I used to think that choosing between `ReentrantLock` and `StampedLock.asWriteLock` <sup>1</sup> is trading reentrancy + optional fairness for lower overhead + the ability to unlock from an arbitrary thread (this one may be necessary in asynchronous code in rare situations), so I was defaulting to `StampedLock.asWriteLock` unless I need either reentrancy or fairness.

Turned out (I measured with OpenJDK JDK 20), `ReentrantLock` and `StampedLock.asWriteLock` have virtually identical throughput both when the critical section is empty and when it involves a bit of computations. Thus, actually, the trade is between reentrancy + optional fairness and the ability to unlock from an arbitrary thread. Since the former is needed much more often, I no longer default to `StampedLock.asWriteLock`, but there is still code where I used it just because I defaulted to it (like `QueryBatchCursor`). There is nothing wrong with that, but there is no reason to preserve `StampedLock.asWriteLock` if it introduces any inconvenience.

---

<sup>1</sup> `StampedLock` is most famous for its optimistic reads functionality, which involves no locking, but this is irrelevant to us here.

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java [hunk 10]
```diff
@@ -193,39 +191,44 @@ public int getMaxWireVersion() {
...
-                                    if (nextBatch.isEmpty() && commandCursorResult.getServerCursor() != null) {
-                                        getMore(commandCursorResult.getServerCursor(), wrappedCallback);
-                                    } else {
-                                        wrappedCallback.onResult(nextBatch, null);
-                                    }
-                                }), callback);
+                getMoreLoop(assertNotNull(connection), cursor, wrappedCallback), callback);
+    }
+
+    private void getMoreLoop(final AsyncConnection connection, final ServerCursor serverCursor,
```
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:197]: I wonder if using `AsyncCallbackLoop` will make the code simpler (it will remove the `getMoreLoop` recursion, but that does not necessary mean that the code will be simpler). I will play with the code and see how it is.

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java]: Updated to use RentrantLock in the base class.

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java [hunk 11]
```diff
@@ -102,120 +83,99 @@ class AsyncCommandBatchCursor<T> implements AsyncAggregateResponseBatchCursor<T>
...
-                    this.pinnedConnection = connection.retain();
-                    this.pinnedConnection.markAsPinned(Connection.PinningMode.CURSOR);
-                }
-            }
+        this.maxWireVersion = connectionDescription.getMaxWireVersion();
+        this.firstBatchEmpty = commandCursorResult.getResults().isEmpty();
+
+        AsyncConnection connectionToPin = null;
+        boolean releaseServerAndResources = false;
+        if (limitReached()) {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:99]: If simplifying this code helps, it's safe to remove all the `limitReached` checks.  The limit checks are a remnant of when the driver supported OP_QUERY as well as the `find` command for implementing the find helper.   OP_QUERY required the client to enforce the limit, while with the `find` command the server enforces the limit and when it is reached will return the last batch with a cursorId of 0.

Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:99]: Ack - will remove and solely rely on the server

Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:99]: The batchSize calculation for getMore is also unnecessary, but there are spec tests that still require it.

Comment (stIncMale) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:99]: > it's safe to remove all the limitReached checks

Is this also true for the sync cursor?

Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:99]: Yes, it's a property of the find command itself, not sync/async distinction

Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java]: Done.

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java [hunk 12]
```diff
@@ -19,203 +19,150 @@
...
-                    return true;
-                });
-                if (doGetMore) {
-                    getMore(localCursor, callback);
-                }
+                getMore(localServerCursor, funcCallback);
             }
-        }
+        }).get((r, t) -> {
+            if (resourceManager.getServerCursor() == null) {
```
Comment (rozza) [inline driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:122]: I think we can remove the auto-closing nature of these AsyncBatchCursors now.

I propose adding a `hasNext` method which just then returns `resourceManager.getServerCursor() != null` for async.

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java (hunk 11 above)
Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java]: > The batchSize calculation for getMore is also unnecessary, but there are spec tests that still require it.

With the latest changes, the `A successful find event with a getmore and the server kills the cursor (<= 4.4)` fails due to the change in how batchSize is set for getMore.  You'll either have to roll back the part of the change that sets the batchSize, or change the test (this is the subject of DRIVERS-1448).

**Comment by jyemin** [comment]
The `ChangeStreamBatchCursorSpecification#should call the underlying CommandBatchCursor` is failing for me locally with

```
com.mongodb.MongoException
	at app//com.mongodb.MongoException.fromThrowableNonNull(MongoException.java:84)
	at app//com.mongodb.internal.operation.ChangeStreamBatchCursor.resumeableOperation(ChangeStreamBatchCursor.java:192)
	at app//com.mongodb.internal.operation.ChangeStreamBatchCursor.next(ChangeStreamBatchCursor.java:79)
	at com.mongodb.internal.operation.ChangeStreamBatchCursorSpecification.should call the underlying CommandBatchCursor(ChangeStreamBatchCursorSpecification.groovy:49)
Caused by: java.lang.AssertionError
	at com.mongodb.assertions.Assertions.assertNotNull(Assertions.java:181)
	at com.mongodb.internal.operation.ChangeStreamBatchCursor.convertAndProduceLastId(ChangeStreamBatchCursor.java:174)
	at com.mongodb.internal.operation.ChangeStreamBatchCursor.lambda$next$2(ChangeStreamBatchCursor.java:81)
	at com.mongodb.internal.operation.ChangeStreamBatchCursor.resumeableOperation(ChangeStreamBatchCursor.java:189)
	... 2 more

```

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java (hunk 10 above)
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:183]: No, it hardly reduces the amount of code, and makes producing a result from a loop more difficult (requires additional code to do that), so the change is not beneficial here.

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java (hunk 11 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java]: Thank you - updated the test(s)

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java [hunk 13]
```diff
@@ -95,17 +91,11 @@ class AsyncCommandBatchCursor<T> implements AsyncAggregateResponseBatchCursor<T>
         this.firstBatchEmpty = commandCursorResult.getResults().isEmpty();
 
         AsyncConnection connectionToPin = null;
-        boolean releaseServerAndResources = false;
-        if (limitReached()) {
-            releaseServerAndResources = true;
-        } else if (connectionDescription.getServerType() == ServerType.LOAD_BALANCER) {
+        if (connectionDescription.getServerType() == ServerType.LOAD_BALANCER) {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:94]: Nit: Now that this is a single `if`, it can become a simple assignment with a ternary expression.

File: driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursor.java [hunk 14]
```diff
@@ -93,17 +90,11 @@ class CommandBatchCursor<T> implements AggregateResponseBatchCursor<T> {
         this.firstBatchEmpty = commandCursorResult.getResults().isEmpty();
 
         Connection connectionToPin = null;
-        boolean releaseServerAndResources = false;
-        if (limitReached()) {
-            releaseServerAndResources = true;
-        } else if (connectionDescription.getServerType() == ServerType.LOAD_BALANCER) {
+        if (connectionDescription.getServerType() == ServerType.LOAD_BALANCER) {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursor.java:93]: As above: make it a ternary expression at the site of the variable declaration

File: driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursor.java [hunk 15]
```diff
@@ -263,9 +246,6 @@ private void getMore() {
                 throw translateCommandException(e, serverCursor);
             }
             resourceManager.setServerCursor(nextServerCursor);
-            if (limitReached() || !resourceManager.operable()) {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursor.java:278]: Does it mean that if the resource manager is in the `CLOSE_PENDING` state after a call to `#close`, then the resources will never be released?  If so, I'm not sure this is a desirable behavior change.

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java (hunk 12 above)
Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:122]: I commented on what I think is the same change in the sync version, where it's easier to understand.

**Review by jyemin — CHANGES_REQUESTED** [review]

File: driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursor.java (hunk 15 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursor.java:278]: The `!resourceManager.operable()` check here wasn't needed as @stIncMale proved in mongodb/mongo-java-driver#1230 (JAVA-5124). So there wasn't a bug just a bad test - which was also updated as part of this change.

TLDR; endOperation handles closing / releasing of resources.

Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursor.java:278]: I see.  Thanks

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java (hunk 12 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:122]: This is different to the sync issue (or I missed which one was commented on). AsyncBatchCursors automatically mark themselves as closed on consumption of all resources. That predates these refactorings.

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java (hunk 13 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java]: Done

File: driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursor.java (hunk 14 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/CommandBatchCursor.java]: Done

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java (hunk 12 above)
Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:118]: I think my comment on the sync version is about something different, so ignore my previous comment on this thread.

Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:118]: I _think_ what you're saying is that users of the batch cursors will now have to call `close` explicitly in order to release resources (i.e the pinned connection and the connection source).  If so, that is a behavioral change that we have to consider carefully, as it effects not just driver internals but users of the driver.  Consider what we've allowed in the legacy API:

```
DBCursor cursor = dbCollection.find();
while (cursor.hasNext()) {
   cursor.next();
}
// allow cursor to go out of scope and let the cursor cleaner call DBCursor#close, which in turn calls CommandBatchCursor#close
```

Currently, if the above loop exits normally (without throwing), all the resources will be released and the cursor cleaner is a no-op.  With this proposed change, the releases won't be released until GC kicks in.

While the above is shoddy application coding, we have allowed it and I'm hesitant to change its behavior unless there is a compelling reason.

File: driver-core/src/main/com/mongodb/internal/operation/CursorResourceManager.java [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+        return state;
+    }
+
+    @Nullable
+    public CS getConnectionSource() {
+        return connectionSource;
+    }
+
+    @Nullable
+    public C getPinnedConnection() {
```
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/internal/operation/CursorResourceManager.java:103]:
```suggestion
    /**
     * Thread-safe.
     */
    @Nullable
    C getPinnedConnection() {
```

File: driver-core/src/main/com/mongodb/internal/operation/CursorResourceManager.java [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+        return pinnedConnection;
+    }
+
+    public boolean isSkipReleasingServerResourcesOnClose() {
```
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/internal/operation/CursorResourceManager.java:107]:
```suggestion
    /**
     * Thread-safe.
     */
    boolean isSkipReleasingServerResourcesOnClose() {
```

**Review by stIncMale — CHANGES_REQUESTED** [review]
Reviewed `CommandBatchCursor` and `CursorResourceManager` with regard to how it is used in the synchronous code.

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java (hunk 12 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:118]: Ah that makes more sense, let me capture that in a test and apply the fix to release.

Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:118]: nm - there are tests that already cover this scenario for example: 
> should handle getMore when there are empty results but there is a cursor

So there is no change in behaviour here - because setting the server cursor to null `resourceManger.setServerCursor` already releases the resources automatically if the cursor is null.

File: driver-core/src/main/com/mongodb/internal/operation/CursorResourceManager.java [hunk 18]
```diff
@@ -0,0 +1,263 @@
...
+            final MongoNamespace namespace,
+            final CS connectionSource,
+            @Nullable final C connectionToPin,
+            @Nullable final ServerCursor serverCursor) {
+        this.lock = new ReentrantLock();
+        this.namespace = notNull("mongoNamespace", namespace);
+        this.state = State.IDLE;
+        if (serverCursor != null) {
+            connectionSource.retain();
+            this.connectionSource = notNull("connectionSource", connectionSource);
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/CursorResourceManager.java]: Removed unnecessary checks

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java (hunk 12 above)
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:117]: I agree, that the code
```java
if (getServerCursor() == null) {
    close();
}
```
is not needed for releasing resources (and there is no such code in `CommandBatchCursor`), but it is needed to self-close the `AsyncCommandCatchCursor`. As far as I understand, without this code the `AsyncCommandCatchCursor.isClosed` will continue returning `false` despite all resources having been released (`CommandBatchCursor` does not have this issue because it does not have the `isClosed` method).

Thus, I think we should either live the code as is, and add a comment explaining what's going on.

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java [hunk 19]
```diff
@@ -228,187 +170,149 @@ public int getMaxWireVersion() {
...
-            document.append("maxTimeMS", new BsonInt64(maxTimeMS));
+
+        /**
+         * Thread-safe.
+         * Executes {@code operation} within the {@link #tryStartOperation()}/{@link #endOperation()} bounds.
+         */
+        <R> AsyncCallbackSupplier<R> execute(final AsyncCallbackSupplier<R> callbackSupplier) {
+            return callback -> {
+                if (!tryStartOperation()) {
+                    callback.onResult(null, new IllegalStateException(MESSAGE_IF_CONCURRENT_OPERATION));
```
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:239]: `tryStartOperation` returns `false` iff the cursor is closed, which means that `MESSAGE_IF_CONCURRENT_OPERATION` is improper here. `CommandBatchCursor` uses `MESSAGE_IF_CLOSED_AS_CURSOR` in this situation.

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java [hunk 20]
```diff
@@ -19,203 +19,145 @@
...
-            callback.onResult(null, new MongoException("next() called after the cursor was closed."));
-        } else if (initialCommandCursorResult != null && !firstBatchEmpty) {
-            // May be empty for a tailable cursor
-            List<T> results = initialCommandCursorResult.getResults();
-            initialCommandCursorResult = null;
-            if (getServerCursor() == null) {
-                close();
+            callback.onResult(null, new MongoException(MESSAGE_IF_CLOSED_AS_CURSOR));
+            return;
+        }
```
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:101]: `tryStartOperation` in `execute` already handles the situation when `AsyncCommandBatchCursor` is closed in a non-racy way. Let's remove this code.

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java [hunk 21]
```diff
@@ -228,187 +170,149 @@ public int getMaxWireVersion() {
...
+            super(namespace, connectionSource, connectionToPin, serverCursor);
         }
-        if (maxTimeMS != 0) {
-            document.append("maxTimeMS", new BsonInt64(maxTimeMS));
+
+        /**
+         * Thread-safe.
+         * Executes {@code operation} within the {@link #tryStartOperation()}/{@link #endOperation()} bounds.
+         */
+        <R> AsyncCallbackSupplier<R> execute(final AsyncCallbackSupplier<R> callbackSupplier) {
```
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:236]: This method is named `execute`, and is documented as `Executes {@code operation} ...`, but actually it does not execute anything, it just creates a new `AsyncCallbackSupplier` based on the given `callbackSupplier`, which is actually the `operation`. For it to stay true to its documentation and for the code to be similar to `execute` in `CommandBatchCursor`, the method should guard and actually asynchronously execute the operation it is given.

I expressed the proposed code changes in https://github.com/stIncMale/mongo-java-driver/commit/97ebded4ba552ceb569e66e4c73bdf8e2069d27e (haven't run any tests).

File: driver-core/src/main/com/mongodb/internal/operation/CursorResourceManager.java [hunk 22]
```diff
@@ -0,0 +1,277 @@
...
+    boolean tryStartOperation() throws IllegalStateException {
+        return withLock(lock, () -> {
+            State localState = state;
+            if (!localState.operable()) {
+                return false;
+            } else if (localState == State.IDLE) {
+                state = State.OPERATION_IN_PROGRESS;
+                return true;
+            } else if (localState == State.OPERATION_IN_PROGRESS) {
+                throw new IllegalStateException(MESSAGE_IF_CONCURRENT_OPERATION);
```
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/internal/operation/CursorResourceManager.java:151]: This throwing does not work for the asynchronous code in `AsyncCommandBatchCursor.ResourceManager.execute`. A way to fix this is to refactor `tryStartOperation` such that it returns `Optional`/`null`-able `IllegalStateException` either when the state is not operable (the cursor was closed)<sup>1</sup> or when another operation is in progress. Then both `AsyncCommandBatchCursor` and `CommandBatchCursor` may inspect the returned value and propagate the exception using the appropriate mechanism (`throw` in the sync code, `onResult` in the async callback-based code).

P.S. One may argue that `AsyncCommandBatchCursor.next` can't actually be called concurrently with itself (there are no other operations guarded by `execute` on `AsyncCommandBatchCursor`) by a user, because we wrap it in a `Publisher`. So if we observe `localState == State.OPERATION_IN_PROGRESS` here from `AsyncCommandBatchCursor.ResourceManager.execute`, then we know there is a driver bug, i.e., we should rather have an assertion. I think, it is also a valid approach, but in this case `CursorResourceManager.tryStartOperation` has to differentiate when it's called from `CommandBatchCursor.ResourceManager.execute`, and when from `AsyncCommandBatchCursor.ResourceManager.execute`.

---

<sup>1</sup> This also requires the method to start accepting the `exceptionMessageIfClosed` parameter, similarly to `CommandBatchCursor.ResourceManager.execute`.

**Review by stIncMale — CHANGES_REQUESTED** [review]
Partially reviewed `AsyncCommandBatchCursor`.

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java (hunk 12 above)
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:117]: Even better, instead of doing
```
if (getServerCursor() == null) {
    close();
}
```
after `endOperation` in `AsyncCommandBatchCursor`, let's just call `close` instead of `releaseClientResources` in `CursorResourceManager.setServerCursor`. This will have the same effect, and `AsyncCommandBatchCursor` won't need the aforementioned additional code that we don't have in `CommandBatchCursor`.

File: driver-core/src/main/com/mongodb/internal/operation/CursorResourceManager.java [hunk 23]
```diff
@@ -0,0 +1,277 @@
...
+    }
+
+    void setServerCursor(@Nullable final ServerCursor serverCursor) {
+        assertTrue(state.inProgress());
+        assertNotNull(this.serverCursor);
+        // without `connectionSource` we will not be able to kill `serverCursor` later
+        assertNotNull(connectionSource);
+        this.serverCursor = serverCursor;
+        if (serverCursor == null) {
+            releaseClientResources();
```
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/internal/operation/CursorResourceManager.java:230]:
```suggestion
            close();
```

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java (hunk 20 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java]: Ack

File: driver-core/src/main/com/mongodb/internal/operation/CursorResourceManager.java (hunk 23 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/CursorResourceManager.java:230]: Had to revert as cannot do it for both sync and async cursors.

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java (hunk 21 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java]: Cherry-picked and adapted 👍

File: driver-core/src/main/com/mongodb/internal/operation/CursorResourceManager.java (hunk 22 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/CursorResourceManager.java:151]: Opted to throw an assertion error (added `Assertions.doesNotThrow`).

**Review by stIncMale — APPROVED** [review]
There is an optional suggestion, otherwise approved.
